### ISSUE TITLE
Replicable database schema

### ISSUE THREAD

**Issue by IagoSRL** [loconomics/loconomics-data#3]
Per conversations with @ndarilek @samwa @joshdanielson, we need to have a replicable database through the SQL scripts we store and manage at this repo.

This topic was covered in the past with Matt (late 2016!) and we have record of it at [an issue and comments on the main repo](https://github.com/loconomics/loconomics/issues/144#issuecomment-277490125), I'm copying from there the main points of tasks/targets that I think we should cover:

------------
- we need to being able to re-create the whole database from a set of SQL scripts
- and too to track changes (scheme and objects)
- being able to track system tables data and copy it on a database; both data that is tight to code (statuses, types) and tables with manually edited data needed for the system to work (jobtitles, pricingtypes, countries). -> moved to #5
- some of that called system tables (another name?) can receive data introduced by users, that at some points (for testing or reproducing bugs) we may want to copy back from live to dev databases and track changes on scripts at repo.
- this must enable us to automate the creating of a new local database with base data, run tests on it for automated testing; or perform changes to try ideas, branch specific changes, without breaking anyone else.
-----------

That initial proposal of tasks covered both schema and data. **Let's start here with the schema only**, a task already started by @ndarilek by generating a schema file from the *dev* database, that is the "core" of this repo.

**Comment by IagoSRL** [comment]
I pushed branch named is3-replicable-schema-vsproj with the content generated from Visual Studio using the "SQL Project" set-up and run against the *dev* database. Aside from some settings files, it generates a folder named 'dbo' containing database objects DDL scripts per file and folder type.

Visual Studio includes an interesting feature here to compare the schema in the files against a database, displaying a diff of changes and generating the SQL scripts to apply the changes. It let you choose a new database so it creates a full database script.
I don't want to depend on this since is Windows only and because I prefer to be able to automate the task, so looked for alternatives.

Microsoft is making a cross browser solution to manage SQL Server and Azure Databases, built as a fork of VSCode and named ["Azure Data Studio"](https://cloudblogs.microsoft.com/sqlserver/2018/10/18/the-october-release-of-azure-data-studio-is-now-available/) (was named SQL Operations Studio initially). It's meant more for developing and analyzing data than server management as of now, including support to manage projects like in a code editor, so I have hope it could include that VS feature but it doesn't. Same with extensions for VSCode or command line tools.

I'm thinking that for the task of just recreate a database from scratch, we may find a safety way to concatenate the files, disabling constraints at the beginning and enabling then at the end so the script runs without intermediate errors caused by wrong sort of dependencies. I did some of that temporary constraint disabling for smaller scripts, will try.

**Comment by IagoSRL** [comment]
Important progress here!

I was able to build a simple tool (a Nodejs script) at this repo, that creates the `schema.sql` file automatically using the individual file scripts. This way, we should consider the sql files under `./schema` the only **source of truth** about the database.

I had to to some changes on the scripts to make the tool and generated execute correctly. Special care for tables, where I had to move declarations of foreign key constraints out of the `CREATE TABLE` statement, separated by a batch separator (`GO`) in a new `ALTER TABLE` statement. This way, the tool can place both parts of the script (before and after the first batch separator) at the proper place in the final script to prevent dependencies errors.
Too, I added named constraints to some 'anonymous' FKs, I keep a list of them so I can later udpate them at the dev and live databases. That's prevents some errors when applying updates to database because of the name being autogenerated on each one.

Right now, this changes are in the *master* branch, and the last `schema.sql` file there was generated by the tool (just run `yarn build-schema-sql`).
I updated the README about all this things.

By running the new generated `schema.sql`, I was able to successfully execute it without errors. Just some warnings, all the same type with a message like `The module 'TestAllUserAlerts' depends on the missing object 'TestProfileActivation'. The module will still be created; however, it cannot run successfully until the object exists.`. It's a warning, so nothing stop being created because of that, and if the objects referenced actually exists they will end working correctly. And any proc, view or function showing this warnings used in production are actually fine; may be some that are unused and unmaintained for a while where the warning is a real error, that ones needs to be removed as part of a clean-up. Using the Visual Studio SQL project I can locate them because has a linting tool integrated that highlights real problems like that.
Other output lines are just notices of objects being created.

Anyway, nothing that stop us from using this schema and trust and rely on this source.

Let me know if you have some problem or doubt.
